Implementing chart report initial structure

mail2bishnoi opened the pull request:
Hi Anton,

I started working on chart reports. Although it is in a very early stage, a few points I would like to discuss with you -

1. Since the chart type attribute does not go into the input of the API, I did not include it in the report filter form but added it to the chart component itself.
2. Which is the best way to convert a nested position object to a simple object? Because we need to draw a graph with attributes like distance, totalDistance, etc

![chart-report](https://user-images.githubusercontent.com/41992346/99779102-9605eb80-2b3a-11eb-8fba-363b64b41b96.PNG)

![chart-report-2](https://user-images.githubusercontent.com/41992346/99779118-9dc59000-2b3a-11eb-8928-666af69f3b25.PNG)

Best Regards

tananaev commented on modern/src/common/formatter.js:
```diff
@@ -9,7 +9,7 @@ export const formatPosition = (value, key) => {
     case 'deviceTime':
     case 'serverTime':
-      return moment(value).format('LLL');
+      return moment(value).format('YYYY-MM-DD HH:mm');
```
This is not a universal solution. We need a better one.

mail2bishnoi replied:
Changes reverted. Tried a new solution.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,74 @@
+
+import {LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend, ResponsiveContainer } from 'recharts';
+
+const useStyles = makeStyles((theme) => ({
```
Why do we need this?

mail2bishnoi replied:
Sorry. Forgot to remove it during code cleanup.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,74 @@
+        if (contentType === 'application/json') {
+          setItems(await response.json());
+        } else {
+          window.location.assign(window.URL.createObjectURL(await response.blob()));
```
Why is this here? Please make sure you clean up all unnecessary stuff.

mail2bishnoi replied:
Cleaned up.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,74 @@
+        <CardHeader action={<ChartType type={type} setType={setType}/>} />
+        <Divider />
+        <CardContent>
+          <Box height={400} position="relative">
```
Looks like you have a lot of hardcoded styles. You have to clean up all that.

mail2bishnoi replied:
I'll do some workaround.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,74 @@
+const CustomizedAxisTick = ({ x, y, stroke, payload }) =>{
+  return (
+    <g transform={`translate(${x},${y})`}>
+      <text x={0} y={0} dy={16} textAnchor="end" fill="#666" transform="rotate(-35)">{payload.value}</text>
```
Please explain what this is for.

mail2bishnoi replied:
To show the custom X-axis lable.

tananaev commented on modern/src/reports/ChartType.js:
```diff
@@ -0,0 +1,53 @@
+
+  return (
+    <FormControl className={classes.formControl}>
+      <Select value={type} onChange={e => setType(e.target.value)} >
```
I think we should start with a smaller subset for now. I don't think it's a good way to maintain this list specifically for charts. Maybe we can have a list somewhere else as a more generic list that can be converted into select.

mail2bishnoi replied:
Removed this component and added chartTypes list and then created a select using chartTypes.

tananaev commented:
Added some comments. As for your questions.

> Since the chart type attribute does not go into the input of the API, I did not include it in the report filter form but added it to the chart component itself.

I think it should still live on the left side. Maybe a separate card below the main one.

> Which is the best way to convert a nested position object to a simple object? Because we need to draw a graph with attributes like distance, totalDistance, etc

How does it work now? Can't you just use a map function?

mail2bishnoi commented:
Hi Anton,

I have tried to incorporate your suggestions. I have shifted the chart type to the left side.
With this, I have tried to format the position. Please review once and give suggestions for the improvements.

An issue related to the chart library is open due to which the problem can sometimes occur in the chart, such as the line going out of the chart boundary.

[https://github.com/recharts/recharts/issues/2149](url)

Best Regards

tananaev commented:
Can you please provide an updated screenshot.

mail2bishnoi commented:
PFA
1. Char with no issues.
![chart1](https://user-images.githubusercontent.com/41992346/100185015-c964c380-2f08-11eb-814d-2d89e5874d8b.PNG)


2. Chart with the issue in recharts (line going outside chart boundary).
![chart2](https://user-images.githubusercontent.com/41992346/100185131-18125d80-2f09-11eb-9b06-5eb730351c7e.PNG)

tananaev commented:
I was thinking of something like this:
<img width="257" alt="Screen Shot 2020-11-24 at 9 06 30 PM" src="https://user-images.githubusercontent.com/1161863/100185683-46585480-2e99-11eb-8de9-4acf483a6194.png">

mail2bishnoi commented:
It can also be used.

Is the rest of the approach correct?

tananaev commented:
I think it looks good overall. There are probably some minor comments that we can once you have main structure ready.

The bug in the library is a bit worrying. Are there any other better libraries?

mail2bishnoi commented:
Ok. Thanks.

I'll look for some other chart library and update you.

mail2bishnoi commented:
Hi Anton,

Incorporated your suggestions. Please review and suggest improvements.

with this, I have fixed the chart issue by converting speed into the number.

![chart3](https://user-images.githubusercontent.com/41992346/100203107-3045a500-2f28-11eb-9ceb-34c5218aa329.PNG)

tananaev commented on modern/src/common/chartTypes.js:
```diff
@@ -0,0 +1,7 @@
+import t from '../common/localization';
+
+export const chartTypes = [
```
Let's inline this for now in the charts module. I think longer term we want something more general to list all known position attributes.

mail2bishnoi replied:
done.

tananaev commented on modern/src/common/formatter.js:
```diff
@@ -80,3 +80,26 @@ export const formatVolume = (value, unit) => {
   return moment.duration(value).humanize();
 };
+
+const speedConverter = (value, unit) => {
```
There's too much duplication with the functions above. Need to unify that. And also I think we should move converter functions into a separate file.

mail2bishnoi replied:
done.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,68 @@
+      const data = await response.json();
+      let formattedData = data.map((obj)=>{
+        return Object.assign(obj, 
+                              {speed: Number(getConverter('speed')(obj.speed, speedUnit))},
```
Fix formatting here.

mail2bishnoi replied:
done.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,68 @@
+        <Box height={400}>
+          <ResponsiveContainer>
+            <LineChart data={items}>
+              <XAxis dataKey="fixTime" tick={<CustomizedAxisTick/>} height={60} />
```
There are some hardcoded numbers like height and grid. Is there a way to get rid of them? Can you please clarify what each one is for.

mail2bishnoi replied:
Box->height is for setting chart height. I have tried setting the chart height using CSS styling but it's not working.

rest of the hardcoded style is chart specific. Xaxis->height is for setting x-axis height so that date and time can be rotated and still visible.

tananaev commented on modern/src/reports/ReportLayoutPage.js:
```diff
@@ -29,6 +35,18 @@ const ReportLayoutPage = ({ reportFilterForm:ReportFilterForm, setItems, childre
+                <FormControl variant="filled" margin="normal" fullWidth>
+                <InputLabel>{t('reportChartType')}</InputLabel>
+                <Select defaultValue="speed" onChange={e => setType(e.target.value)}>
+                {chartTypes.map(item => (
```
I think we need indentation here.

mail2bishnoi replied:
done

tananaev commented on modern/src/reports/ReportLayoutPage.js:
```diff
@@ -29,6 +35,18 @@ const ReportLayoutPage = ({ reportFilterForm:ReportFilterForm, setItems, childre
+              <Paper className={classes.chart}>
+                <FormControl variant="filled" margin="normal" fullWidth>
+                <InputLabel>{t('reportChartType')}</InputLabel>
+                <Select defaultValue="speed" onChange={e => setType(e.target.value)}>
```
Looks like we define this default value in state and here. Can we avoid that?

mail2bishnoi replied:
done

tananaev commented on modern/src/common/converter.js:
```diff
@@ -0,0 +1,24 @@
+  return (value * factor).toFixed(2);
+};
+
+export const getConverter = (key) => {
```
I don't think we need this. Please just export separate functions instead.

mail2bishnoi replied:
done

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,73 @@
+      const data = await response.json();
+
+      let formattedData = data.map(obj => {
+        return Object.assign(
```
I don't like reusing the same model. Let's create a separate new object with only the data that we need.

mail2bishnoi replied:
done

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,73 @@
+
+      const data = await response.json();
+
+      let formattedData = data.map(obj => {
```
nit Please don't use shortened words. You can use `it` or `position` or `item` here instead.

mail2bishnoi replied:
done

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,73 @@
+      let formattedData = data.map(obj => {
+        return Object.assign(
+          obj, 
+          {speed: Number(getConverter('speed')(obj.speed, speedUnit))},
```
Formatting is still off here.

tananaev commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,74 @@
+
+      let formattedPositions = positions.map(position => {
+        return {
+          speed:Number(speedConverter(position.speed, speedUnit)),
```
Can you please make sure formatting matches existing code. You need a space after `:`.

mail2bishnoi replied:
done.

tananaev commented on modern/src/reports/ReportLayoutPage.js:
```diff
@@ -29,6 +34,18 @@ const ReportLayoutPage = ({ reportFilterForm:ReportFilterForm, setItems, childre
             <Paper className={classes.form}>
               <ReportFilterForm setItems={ setItems } />
             </Paper>
+            {showChartType && (
```
I just noticed one big thing here. I don't think it makes sense to include chart here. It's used only in the chart report.

mail2bishnoi replied:
What would you like to suggest where it should be shifted?

Should I move it again to the right side as before?

tananaev replied:
Maybe we need to split layout differently somehow to allow it?

mail2bishnoi replied:
Hi Anton,

I am unable to understand how to implement this? Can you give me a direction?

One solution I thought that includes the Report Filer component and Chart component in a single ChartReportPage.

Thanks

tananaev replied:
Maybe you can pass filter using props. Something like this:
```js
        filter={
          <ReportFilterForm setItems={ setItems } />
          <Paper className={classes.chart}>
            <FormControl variant="filled" margin="normal" fullWidth>
            ...
        }
```

mail2bishnoi replied:
Done. Please check and suggest improvements.

mail2bishnoi commented:
Hi Anton,

Any updates on this?

tananaev commented:
Sorry, I added comments, but forgot to submit the review.

mail2bishnoi commented on modern/src/reports/ReportLayoutPage.js:
```diff
@@ -27,7 +27,7 @@ const ReportLayoutPage = ({ reportFilterForm:ReportFilterForm, setItems, childre
           <Grid item xs={12} md={3} lg={2}>
             <Paper className={classes.form}>
-              <ReportFilterForm setItems={ setItems } />
+              <ReportFilterForm setItems={ setItems } type={ type } setType={ setType }/>
```
Sorry dear. Just forgot.

mail2bishnoi commented on modern/src/reports/ChartReportPage.js:
```diff
@@ -0,0 +1,85 @@
+  const [type, setType] = useState('speed');
+
+  return (
+    <ReportLayoutPage reportFilterForm={ReportFilterForm} setItems={setItems} type={type} setType={setType}>
```
Chart type

mail2bishnoi commented:
Hi Anton,

Your suggestion included.

Thank you so much for providing feedback and teaching me how to code on a production-level project.

tananaev commented on modern/src/reports/ReportLayoutPage.js:
```diff
@@ -18,7 +18,7 @@ const useStyles = makeStyles(theme => ({
 }));
 
-const ReportLayoutPage = ({ reportFilterForm:ReportFilterForm, setItems, children }) => {
+const ReportLayoutPage = ({ reportFilterForm:ReportFilterForm, setItems, type, setType, children }) => {
```
I don't like that we need to pass through a parameter that is not generic. We need to fix that somehow. Maybe component breakdown should be different.

mail2bishnoi commented:
Hi Anton,

Didn't find any generic solution to the above problem.


Can I use the below solution for passing data and chart type?
`const [items, setItems]=useState({data:[], type:'speed'})`;

tananaev commented:
Instead of passing a component name to `reportFilterForm` can we pass the whole JSX?

mail2bishnoi commented:
Hi Anton,

This time I have tried to refactor all the report components to make them more generic. I believe I did it in the best way.

Please review the code and suggest improvements.

Best Regards

tananaev commented:
Merged. Thanks for the hard work. It looks really good now.

pc500 commented:
I have noticed that the "Speed" value in the modern UI does not correctly show MPH.  It appears to be roughly 8% "slow".  You can see the same from the new vs old UI.

![image](https://user-images.githubusercontent.com/33823824/111076878-845ea000-84ab-11eb-83c4-365c8f61e7bb.png)

![image](https://user-images.githubusercontent.com/33823824/111076915-a1936e80-84ab-11eb-8e04-378b20d95ff4.png)
